Tests/OptimizelyTests-DataModel/OptimizelyJSONTests.swift

```diff
@@ -0,0 +1,489 @@
+        }
+        """
+        self.data = [
+            "field1": 1,
+            "field2": 2.5,
+            "field3": "three",
+            "field4": [
+                "inner_field1": 3.0,
+                "inner_field2": ["1","2",3.01,4.23,true]
+            ],
+            "field5": true,
+        ]
+        try! self.optimizelyJSON = OptimizelyJSON(payload: self.payload)
+    }
+}
+
+// MARK: - Tests
+
+extension OptimizelyJSONTests {
+    
+    func testConstructorsHappyPath() {
+        var err: Error?
+        var optimizelyJSON1: OptimizelyJSON?
+        do {
+            try optimizelyJSON1 = OptimizelyJSON(payload: self.payload)
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        
+        var optimizelyJSON2: OptimizelyJSON?
+        do {
+            try optimizelyJSON2 = OptimizelyJSON(data: self.data)
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        
+        var optimizelyJSON1ToString: String?
+        var optimizelyJSON2ToString: String?
+        do {
+            try optimizelyJSON1ToString = optimizelyJSON1?.toString()
+            try optimizelyJSON2ToString = optimizelyJSON2?.toString()
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        XCTAssertNotNil(optimizelyJSON1ToString)
+        XCTAssertEqual(optimizelyJSON1ToString, self.payload)
+        // We cannot compare both string since when converting dict to string, the order is not certain
+        XCTAssertNotNil(optimizelyJSON2ToString)
+        
+        var optimizelyJSON1ToMap: [String:Any]!
+        var optimizelyJSON2ToMap: [String:Any]!
+        do {
+            try optimizelyJSON1ToMap = optimizelyJSON1?.toMap()
+            try optimizelyJSON2ToMap = optimizelyJSON2?.toMap()
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        XCTAssertNotNil(optimizelyJSON1ToMap)
+        XCTAssertNotNil(optimizelyJSON2ToMap)
+        XCTAssertTrue(NSDictionary(dictionary: optimizelyJSON1ToMap).isEqual(to: optimizelyJSON2ToMap))
+        
+        // Verifying json2 toString was valid by converting it to dict and comparing with json1 dict
+        let json2StringToMap = OTUtils.convertToDictionary(text: optimizelyJSON2ToString!)
+        XCTAssertNotNil(json2StringToMap)
+        XCTAssertTrue(NSDictionary(dictionary: optimizelyJSON1ToMap).isEqual(to: json2StringToMap!))
+    }
+    
+    func testConstructorWithInvalidPayloadString() {
+        var err: Error?
+        var optimizelyJSON: OptimizelyJSON?
+        do {
+            try optimizelyJSON = OptimizelyJSON(payload: "incorrect_string")
+        } catch {
+            err = error
+        }
+        XCTAssertEqual(err?.localizedDescription, OptimizelyError.failedToConvertStringToDictionary.reason)
+        XCTAssertNil(optimizelyJSON)
+    }
+    
+    func testConstructorWithInvalidJSONArrayPayloadString() {
+        let jsonArrayPayload = """
```

Comment:
what's wrong here?

Reply:
OptimizelyJSON should only accept `JSON Map` as parameter, `JSON Array` should not be acceptable.